eppo_core/src/attributes.rs

```diff
@@ -1,8 +1,48 @@
-use std::collections::HashMap;
+use std::{collections::HashMap, sync::Arc};
 
 use derive_more::From;
 use serde::{Deserialize, Serialize};
 
+use crate::Str;
+
+/// `Subject` is a bundle of subject attributes and a key.
+#[derive(Debug)]
+pub(crate) struct Subject {
+    /// Subject key encoded as attribute value. Known to be `AttributeValue::String`. This is
+    /// done to allow returning subject key as an attribute when rule references "id".
+    key: AttributeValue,
+    attributes: Arc<Attributes>,
+}
```

Comment:
The `Subject` is here to avoid copying subject attributes to add the implicit "id" attribute.